lib/go/thrift/fast_buffer.go

```diff
@@ -0,0 +1,257 @@
+	if p.frameSize == 0 {
+		p.frameSize, err = p.readFrameHeader()
+		if err != nil {
+			return
+		}
+		_, err = p.readAll(p.frameSize)
+		if err != nil {
+			return
+		}
+	}
+	if p.frameSize < num {
+		return nil, fmt.Errorf("Not enought frame size %d to read %d bytes", p.frameSize, num)
+	}
+	b = p.b[p.rIdx : p.rIdx+num]
+	p.frameSize = p.frameSize - num
+	if p.frameSize < 0 {
+		return nil, fmt.Errorf("Negative frame size")
+	}
+	p.rIdx += num
+	return b, nil
+}
+
+func (p *FastFrameBuffer) ReadI64() (num int64, err error) {
+	if p.frameSize == 0 {
+		p.frameSize, err = p.readFrameHeader()
+		if err != nil {
+			return
+		}
+		_, err = p.readAll(p.frameSize)
+		if err != nil {
+			return
+		}
+
+	}
+	if p.frameSize < 8 {
+		return 0, fmt.Errorf("Not enought frame size %d to read %d bytes", p.frameSize, 2)
+	}
+	num = int64(uint64(p.b[p.rIdx+7]) | uint64(p.b[p.rIdx+6])<<8 | uint64(p.b[p.rIdx+5])<<16 | uint64(p.b[p.rIdx+4])<<24 | uint64(p.b[p.rIdx+3])<<32 | uint64(p.b[p.rIdx+2])<<40 | uint64(p.b[p.rIdx+1])<<48 | uint64(p.b[p.rIdx])<<56)
+	p.frameSize = p.frameSize - 8
+	p.rIdx += 8
+	return num, nil
+}
+
+func (p *FastFrameBuffer) ReadI32() (num int32, err error) {
+	if p.frameSize == 0 {
+		p.frameSize, err = p.readFrameHeader()
+		if err != nil {
+			return
+		}
+		_, err = p.readAll(p.frameSize)
+		if err != nil {
+			return
+		}
+	}
+	if p.frameSize < 4 {
+		return 0, fmt.Errorf("Not enought frame size %d to read %d bytes", p.frameSize, 2)
+	}
+	num = int32(uint32(p.b[p.rIdx+3]) | uint32(p.b[p.rIdx+2])<<8 | uint32(p.b[p.rIdx+1])<<16 | uint32(p.b[p.rIdx])<<24)
+	p.frameSize = p.frameSize - 4
+	p.rIdx += 4
+	return num, nil
+}
+
+func (p *FastFrameBuffer) ReadI16() (num int16, err error) {
+	if p.frameSize == 0 {
+		p.frameSize, err = p.readFrameHeader()
+		if err != nil {
+			return
+		}
+		_, err = p.readAll(p.frameSize)
+		if err != nil {
+			return
+		}
+	}
+	if p.frameSize < 2 {
+		return 0, fmt.Errorf("Not enought frame size %d to read %d bytes", p.frameSize, 2)
+	}
+	num = int16(uint16(p.b[p.rIdx+1]) | uint16(p.b[p.rIdx])<<8)
+	p.frameSize = p.frameSize - 2
+	p.rIdx += 2
+	return num, nil
+}
+
+func (p *FastFrameBuffer) readFrameHeader() (int, error) {
+	p.ResetReader()
+	if _, err := p.w.transport.Read(p.buf[:4]); err != nil {
+		return 0, err
+	}
+	frameSize := int(uint32(p.buf[3]) | uint32(p.buf[2])<<8 | uint32(p.buf[1])<<16 | uint32(p.buf[0])<<24)
+	if frameSize < 0 || frameSize > DEFAULT_MAX_LENGTH {
+		return 0, fmt.Errorf("Incorrect frame size (%d)", frameSize)
+	}
+	return frameSize, nil
+}
+
+// TODO 这个方法 是否会有问题？ block? or?
+func (p *FastFrameBuffer) readAll(num int) (int, error) {
+	// 当容量不足以装下所有数据的时候, 重新申请一下
+	if cap(p.b) < num {
+		rbNew := make([]byte, 2*cap(p.b)+num)
+		copy(rbNew, p.b)
+		p.b = rbNew
+	}
+	var i = 0
+	for i < num {
+		l, err := p.w.transport.Read(p.b[i:])
+		if err != nil {
+			return 0, err
+		}
+		i += l
+	}
+	return i, nil
+}
+
+func str2bytes(s string) []byte {
```

Comment:
At the very minimum, this file needs an `!appengine` build tag and a separate, slower implementation without using `unsafe`, so it can be used in all platforms supported by Go.